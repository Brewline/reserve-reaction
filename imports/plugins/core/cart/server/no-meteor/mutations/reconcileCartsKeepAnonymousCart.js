import ReactionError from "@reactioncommerce/reaction-error";
import { Cart as CartSchema } from "/imports/collections/schemas";
import appEvents from "/imports/node-app/core/util/appEvents";

/**
 * @summary Update account cart to have only the anonymous cart items, delete anonymous
 *   cart, and return updated accountCart.
 * @todo When we add a "save for later" / "wish list" feature, we may want to update this
 *   to move existing account cart items to there.
 * @param {Object} accountCart The account cart document
 * @param {Object} accountCartSelector The MongoDB selector for the account cart
 * @param {Object} anonymousCart The anonymous cart document
 * @param {Object} anonymousCartSelector The MongoDB selector for the anonymous cart
 * @param {MongoDB.Collection} Cart The Cart collection
 * @return {Object} The updated account cart
 */
export default async function reconcileCartsKeepAnonymousCart({
  accountCart,
  accountCartSelector,
  anonymousCart,
  anonymousCartSelector,
  Cart
}) {
  const updatedAt = new Date();

  const modifier = {
    $set: {
      items: anonymousCart.items,
      updatedAt
    }
  };
  CartSchema.validate(modifier, { modifier: true });

  const { modifiedCount } = await Cart.updateOne(accountCartSelector, modifier);
  if (modifiedCount === 0) throw new ReactionError("server-error", "Unable to update cart");

  const updatedCart = {
    ...accountCart,
    items: anonymousCart.items,
    updatedAt
  };

  await appEvents.emit("afterCartUpdate", updatedCart);

  const { deletedCount } = await Cart.deleteOne(anonymousCartSelector);
  if (deletedCount === 0) throw new ReactionError("server-error", "Unable to delete anonymous cart");

  return updatedCart;
}
