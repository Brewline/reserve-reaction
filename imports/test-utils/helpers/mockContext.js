const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  shopId: "FAKE_SHOP_ID",
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "Groups",
  "Inventory",
  "MediaRecords",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "Tags",
  "Templates",
  "Themes",
  "Translations",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
