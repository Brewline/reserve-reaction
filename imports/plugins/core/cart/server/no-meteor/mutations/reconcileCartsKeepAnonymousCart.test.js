import Factory from "/imports/test-utils/helpers/factory";
import mockContext from "/imports/test-utils/helpers/mockContext";
import reconcileCartsKeepAnonymousCart from "./reconcileCartsKeepAnonymousCart";

const { Cart } = mockContext.collections;
const accountId = "accountId";
const accountCart = { _id: "ACCOUNT_CART", accountId };
const accountCartSelector = { accountId };
const anonymousCartSelector = { _id: "123" };
const items = [Factory.CartItem.makeOne()];

test("overwrites account cart items, deletes anonymous cart, and returns updated account cart", async () => {
  const result = await reconcileCartsKeepAnonymousCart({
    accountCart,
    accountCartSelector,
    anonymousCart: {
      items
    },
    anonymousCartSelector,
    Cart
  });

  expect(Cart.deleteOne).toHaveBeenCalledWith(anonymousCartSelector);

  expect(Cart.updateOne).toHaveBeenCalledWith(accountCartSelector, {
    $set: {
      items,
      updatedAt: jasmine.any(Date)
    }
  });

  expect(result).toEqual({
    ...accountCart,
    items,
    updatedAt: jasmine.any(Date)
  });
});

test("throws if deleteOne fails", async () => {
  Cart.deleteOne.mockReturnValueOnce(Promise.resolve({ deletedCount: 0 }));

  const promise = reconcileCartsKeepAnonymousCart({
    accountCart,
    accountCartSelector,
    anonymousCart: {
      items
    },
    anonymousCartSelector,
    Cart
  });

  return expect(promise).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if updateOne fails", async () => {
  Cart.updateOne.mockReturnValueOnce(Promise.resolve({ modifiedCount: 0 }));

  const promise = reconcileCartsKeepAnonymousCart({
    accountCart,
    accountCartSelector,
    anonymousCart: {
      items
    },
    anonymousCartSelector,
    Cart
  });

  return expect(promise).rejects.toThrowErrorMatchingSnapshot();
});
