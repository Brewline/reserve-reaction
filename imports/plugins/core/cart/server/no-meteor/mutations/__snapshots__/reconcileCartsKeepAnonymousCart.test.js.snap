// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if deleteOne fails 1`] = `"Unable to delete anonymous cart"`;

exports[`throws if updateOne fails 1`] = `"Unable to update cart"`;
